// web-audio-test-api/lib/validators/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isString = exports.isPositiveNumber = exports.isPositiveInteger = exports.isNumber = exports.isNullOrInstanceOf = exports.isInteger = exports.isInstanceOf = exports.isFunction = exports.isBoolean = exports.isAudioSource = undefined;

var _isAudioSource2 = require("./isAudioSource");

var _isAudioSource3 = _interopRequireDefault(_isAudioSource2);

var _isBoolean2 = require("./isBoolean");

var _isBoolean3 = _interopRequireDefault(_isBoolean2);

var _isFunction2 = require("./isFunction");

var _isFunction3 = _interopRequireDefault(_isFunction2);

var _isInstanceOf2 = require("./isInstanceOf");

var _isInstanceOf3 = _interopRequireDefault(_isInstanceOf2);

var _isInteger2 = require("./isInteger");

var _isInteger3 = _interopRequireDefault(_isInteger2);

var _isNullOrInstanceOf2 = require("./isNullOrInstanceOf");

var _isNullOrInstanceOf3 = _interopRequireDefault(_isNullOrInstanceOf2);

var _isNumber2 = require("./isNumber");

var _isNumber3 = _interopRequireDefault(_isNumber2);

var _isPositiveInteger2 = require("./isPositiveInteger");

var _isPositiveInteger3 = _interopRequireDefault(_isPositiveInteger2);

var _isPositiveNumber2 = require("./isPositiveNumber");

var _isPositiveNumber3 = _interopRequireDefault(_isPositiveNumber2);

var _isString2 = require("./isString");

var _isString3 = _interopRequireDefault(_isString2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.isAudioSource = _isAudioSource3.default;
exports.isBoolean = _isBoolean3.default;
exports.isFunction = _isFunction3.default;
exports.isInstanceOf = _isInstanceOf3.default;
exports.isInteger = _isInteger3.default;
exports.isNullOrInstanceOf = _isNullOrInstanceOf3.default;
exports.isNumber = _isNumber3.default;
exports.isPositiveInteger = _isPositiveInteger3.default;
exports.isPositiveNumber = _isPositiveNumber3.default;
exports.isString = _isString3.default;
